docs/advanced-topics/reconnecting-mid-game.md

```diff
@@ -0,0 +1,28 @@
+---
+id: reconnecting-mid-game
+title: Reconnecting Mid-Game
+---
+
+In a multiplayer game, it is possible that a client gets disconnected from the server, for example if they have network issues, or if their application or device crashes. When it happens, you might want to give the ability to your players to reconnect to the game.
+
+To allow clients to reconnect in the middle of an active game, there are some considerations to have in mind.
+- If clients have a specific state and some in-game data associated with them and you want them to reconnect with that same state and that same data, you should implement a mechanism to keep that state in memory and associate it to that client when they reconnect (see [Session Management](./session-management.md))
+- If using scene management, keep in mind that connecting to a game launches the synchronization process. If you're using additive scenes, this means that if the client's active main scene is the same as the server's, it will load all additive scenes that are currently loaded on the server, without unloading anything on the client. If that is a behavior you do not want, you can unload those scenes on the client when they disconnect, or before attempting to reconnect, or you can use the `NetworkSceneManager.VerifySceneBeforeLoading` callback to prevent loading scenes that are already loaded on the client.
```

Comment:
- General Guidelines for when a client is disconnected:
  - Preserve the client's current in-game state on both the client and server side
    - This will provide you with any state deltas between the client and server that might have occurred in the time span between when the client was disconnected and when the server detects the client was disconnected.
  - Restore the client's scene hierarchy state as it was just prior to the client having connected for the first time.
    - Typically unloading any additive scenes loaded during the game-session or returning to a "single mode" loaded lobby will accomplish this
- General Guidelines for reconnecting a client:
  - Reconnect the client and wait for the client to be synchronized
      - Once synchronized, use the saved client state to restore the client's state as the last part of synchronization.
- When scene management is enabled:
  - Wait until the client's initial connection synchronization event is triggered before applying the reconnected client's game state.   
  There are three ways to monitor when a client is synchronized:
    - Subscribing to the `NetworkSceneManager.OnSynchronizeComplete` event.
    - Subscribing to the `NetworkSceneManager.OnSceneEvent` and checking for the `SceneEvent.SceneEventType`  to equal `SceneEventType.SynchronizeComplete`.
    - Subscribing to the `NetworkManager.OnClientConnectedCallback`
      - _`The NetworkSceneManager` invokes this when a client's synchronization is complete_
- When scene management is disabled:
  - The same principles apply when scene management is disabled with the exception that you might have additional unique scenarios to take into consideration depending upon the custom implementation.

Reply:
@NoelStephensUnity  I have questions about some of these points:

> * General Guidelines for when a client is disconnected:
>      * Preserve the client's current in-game state on both the client and server side
>          * This will provide you with any state deltas between the client and server that might have occurred in the time span between when the client was disconnected and when the server detects the client was disconnected.

Why do we need those deltas? Shouldn't we only consider the state server-side when it detects the client disconnect?

>      * Restore the client's scene hierarchy state as it was just prior to the client having connected for the first time.
>         * Typically unloading any additive scenes loaded during the game-session or returning to a "single mode" loaded lobby will accomplish this

I can see cases where users might want to keep the scene hierarchy. For example, that is what is done in Boss Room during the automatic reconnect process. Should we mention it here?

Reply:
This is really more of a suggestion... there could be some unique aspects to the users netcode assets that might require them to do some of the "delta checking/adjusting" on the client side.  
Another example might be a game where a player disconnects with an internally stored time that contains their data and perhaps the last "local player health update"...if the player drops and the server doesn't detect this for several seconds then something could have impacted the disconnected player's health (i.e. player couldn't do anything about it) and so the user might add some additional code when it receives what the server "thinks" is the player's health but then the player's client responds with the last health update information and some verification key (delivered when updating)... the server validates the key, compares the timestamp of when it was sent (could be part of the key), and if it is valid restores the player's health server side.

I was just trying to keep a broad stroke brush (i.e. save state when disconnected in all places pertinent...could be only server...could be both server and client).

However, since it was only a suggestion... feel free to delete/remove anything that doesn't make 100% sense to have here.

Reply:
I see. I think in that case that it would make more sense to have it in the session management doc page